### PR TITLE
Clear error message when passphrase not supplied with encrypted private key

### PR DESCRIPTION
**Pull request by lakshmi-kannan**
## Before

```
vagrant@st2test:~$ st2 run core.remote hosts=localhost cmd=date username=vagrant private_key=/home/vagrant/.ssh/test_rsa_pp
.
id: 5763024955fc8c18adf74f00
status: failed
parameters:
  cmd: date
  hosts: localhost
  private_key: '********'
  username: vagrant
result:
  error: "Unable to connect to any one of the hosts: [u'localhost'].

 connect_errors={
  "localhost": {
    "failed": true,
    "traceback": "Traceback (most recent call last):\n  File \"/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2actions/runners/ssh/parallel_ssh.py\", line 241, in _connect\n    client.connect()\n  File \"/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2actions/runners/ssh/paramiko_ssh.py\", line 144, in connect\n    self.client = self._connect(host=self.hostname, socket=self.bastion_socket)\n  File \"/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2actions/runners/ssh/paramiko_ssh.py\", line 578, in _connect\n    client.connect(**conninfo)\n  File \"/opt/stackstorm/st2/local/lib/python2.7/site-packages/paramiko/client.py\", line 380, in connect\n    look_for_keys, gss_auth, gss_kex, gss_deleg_creds, gss_host)\n  File \"/opt/stackstorm/st2/local/lib/python2.7/site-packages/paramiko/client.py\", line 597, in _auth\n    raise saved_exception\nSSHException: not a valid EC private key file\n",
    "timeout": false,
    "succeeded": false,
    "stdout": "",
    "stderr": "",
    "error": "Cannot connect to host. not a valid EC private key file: not a valid EC private key file",
    "return_code": 255
  }
}"
  traceback: "  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2actions/container/base.py", line 89, in _do_run
    runner.pre_run()
  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2actions/runners/ssh/paramiko_ssh_runner.py", line 159, in pre_run
    self._parallel_ssh_client = ParallelSSHClient(**client_kwargs)
  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2actions/runners/ssh/parallel_ssh.py", line 60, in __init__
    connect_results = self.connect(raise_on_any_error=raise_on_any_error)
  File "/opt/stackstorm/st2/local/lib/python2.7/site-packages/st2actions/runners/ssh/parallel_ssh.py", line 90, in connect
    raise NoHostsConnectedToException(msg)
"
vagrant@st2test:~$
```

## After

```
(virtualenv)vagrant@st2dev /m/s/s/st2 ❯❯❯ st2 run core.remote hosts=localhost username=vagrant private_key=/home/vagrant/.ssh/test_rsa_pp cmd=date
.
id: 57633ad7d9d7ed1889ad2098
status: failed
parameters:
  cmd: date
  hosts: localhost
  private_key: /home/vagrant/.ssh/test_rsa_pp
  username: vagrant
result:
  error: "Unable to connect to any one of the hosts: [u'localhost'].

 connect_errors={
  "localhost": {
    "failed": true,
    "traceback": "Traceback (most recent call last):\n  File \"/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/parallel_ssh.py\", line 241, in _connect\n    client.connect()\n  File \"/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/paramiko_ssh.py\", line 144, in connect\n    self.client = self._connect(host=self.hostname, socket=self.bastion_socket)\n  File \"/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/paramiko_ssh.py\", line 549, in _connect\n    raise Exception(msg)\nException: Private key file /home/vagrant/.ssh/test_rsa_pp is passphrase protected. Supply a passphrase!\n",
    "timeout": false,
    "succeeded": false,
    "stdout": "",
    "stderr": "",
    "error": "Failed connecting to host localhost.: Private key file /home/vagrant/.ssh/test_rsa_pp is passphrase protected. Supply a passphrase!",
    "return_code": 255
  }
}"
  traceback: "  File "/mnt/src/storm/st2/st2actions/st2actions/container/base.py", line 89, in _do_run
    runner.pre_run()
  File "/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/paramiko_ssh_runner.py", line 159, in pre_run
    self._parallel_ssh_client = ParallelSSHClient(**client_kwargs)
  File "/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/parallel_ssh.py", line 60, in __init__
    connect_results = self.connect(raise_on_any_error=raise_on_any_error)
  File "/mnt/src/storm/st2/st2actions/st2actions/runners/ssh/parallel_ssh.py", line 90, in connect
    raise NoHostsConnectedToException(msg)
"
(virtualenv)vagrant@st2dev /m/s/s/st2 ❯❯❯
```

### CODE REVIEW

File: st2actions/st2actions/runners/ssh/paramiko_ssh.py [hunk 1]
```diff
@@ -554,6 +554,13 @@ def _connect(self, host, socket=None):
         if self.key_files:
             conninfo['key_filename'] = self.key_files
 
+            try:
+                paramiko.RSAKey.from_private_key_file(self.key_files)
+            except paramiko.ssh_exception.PasswordRequiredException:
+                msg = ('Private key file %s is passphrase protected. Supply a passphrase!' %
+                       self.key_files)
+                raise Exception(msg)
```
Comment (lakshmi-kannan) [inline st2actions/st2actions/runners/ssh/paramiko_ssh.py]: The default message in that exception sends out a cryptic message that's not obvious to user. Hence I send out a generic exception.

File: st2actions/st2actions/runners/ssh/parallel_ssh.py [hunk 2]
```diff
@@ -245,7 +245,6 @@ def _connect(self, host, results, raise_on_any_error=False):
             LOG.exception(error)
             if raise_on_any_error:
                 raise
-            error = ' '.join([self.CONNECT_ERROR, str(ex)])
```
Comment (lakshmi-kannan) [inline st2actions/st2actions/runners/ssh/parallel_ssh.py]: Fixes this 

```
    "error": "Cannot connect to host. Private key file /home/vagrant/.ssh/test_rsa_pp is passphrase protected. Supply a passphrase!: Private key file /home/vagrant/.ssh/test_rsa_pp is passphrase protected. Supply a passphrase!",
```

Note the dupe!

File: st2actions/st2actions/runners/ssh/paramiko_ssh.py (hunk 1 above)
Comment (manasdk) [inline st2actions/st2actions/runners/ssh/paramiko_ssh.py]: Please log the original exception.

File: st2actions/st2actions/runners/ssh/paramiko_ssh.py [hunk 3]
```diff
@@ -554,6 +554,13 @@ def _connect(self, host, socket=None):
         if self.key_files:
             conninfo['key_filename'] = self.key_files
 
+            try:
+                paramiko.RSAKey.from_private_key_file(self.key_files)
+            except paramiko.ssh_exception.PasswordRequiredException:
+                msg = ('Private key file %s is passphrase protected. Supply a passphrase!' %
```
Comment (manasdk) [inline st2actions/st2actions/runners/ssh/paramiko_ssh.py]: Remove the `!`

**Comment by manasdk** [comment]
👍

**Comment by Kami** [comment]
Good catch - thanks!

Let's please also add a test case for it.

**Comment by lakshmi-kannan** [comment]
Added unit test: https://github.com/StackStorm/st2/pull/2755/commits/1bf5009d0a4b6dbedbc9a3da0032079935599941#diff-4e2bf97e5b553c71e4a57637fb0d1253R148

File: st2actions/tests/unit/test_parallel_ssh.py [hunk 4]
```diff
@@ -49,6 +51,8 @@ def test_connect_with_password(self):
             client._hosts_client[host].client.connect.assert_called_once_with(**expected_conn)
 
     @patch('paramiko.SSHClient', Mock)
+    @patch.object(ParamikoSSHClient, '_is_key_file_needs_passphrase',
```
Comment (Kami) [inline st2actions/tests/unit/test_parallel_ssh.py]: Needing to patch this "implementation detail" everywhere is kinda nasty and hard to maintain - is there no other way to achieve that?

Anyways, thanks for adding the test :+1:

**Pull request merged**